pandablocks/utils.py

```diff
@@ -0,0 +1,128 @@
+            expected to be the string representation of a uint32.
+        table_fields_info: The info for tables, containing the number of words per row,
+            and the bit information for fields.
+    Returns:
+        unpacked: A dict containing record information, where keys are field names
+            and values are numpy arrays or a sequence of strings of record values
+            in that column.
+    """
+
+    row_words = table_field_info.row_words
+    data = np.array(words, dtype=np.uint32)
+    # Convert 1-D array into 2-D, one row element per row in the PandA table
+    data = data.reshape(len(data) // row_words, row_words)
+    packed = data.T
+
+    unpacked: Dict[str, UnpackedArray] = {}
+
+    for field_name, field_info in table_field_info.fields.items():
+        offset = field_info.bit_low
+        bit_length = field_info.bit_high - field_info.bit_low + 1
+
+        # The word offset indicates which column this field is in
+        # (column is exactly one 32-bit word)
+        word_offset = offset // 32
+
+        # bit offset is location of field inside the word
+        bit_offset = offset & 0x1F
+
+        # Mask to remove every bit that isn't in the range we want
+        mask = (1 << bit_length) - 1
+
+        value = (packed[word_offset] >> bit_offset) & mask
+
+        if field_info.subtype == "int":
+            # First convert from 2's complement to offset, then add in offset.
+            packing_value = (value ^ (1 << (bit_length - 1))) + (-1 << (bit_length - 1))
+        elif field_info.labels:
+            packing_value = [field_info.labels[x] for x in value]
+        else:
+            packing_value = value
+
+        unpacked.update({field_name: packing_value})
+
+    return unpacked
+
+
+def table_to_words(
+    table: Dict[str, Iterable], table_field_info: TableFieldInfo
+) -> List[str]:
+    """Convert records based on the field definitions into the format PandA expects
+    for table writes.
+
+    Args:
+        table: A dict containing record information, where keys are field names
+            and values are iterables of record values in that column.
+        table_field_info: The info for tables, containing the dict `fields` for
+            information on each field, and the number of words per row.
+    Returns:
+        List[str]: The list of data ready to be sent to PandA
+    """
+    row_words = table_field_info.row_words
+
+    # Iterate over the zipped fields and their associated records to construct the
+    # packed array.
+    packed = None
+
+    for column_name, column in table.items():
+        field_details = table_field_info.fields[column_name]
+        if field_details.labels:
+            # Must convert the list of ints into strings
+            column = [field_details.labels.index(x) for x in column]
+
+        # PandA always handles tables in uint32 format
+        column_value = np.uint32(np.array(column))
```

Comment:
Concerning the discussion of this cast:

Turns out it is necessary:
``` python
x = np.array([1, 2, 3])
y = np.uint32(x)
z = np.zeros(3, dtype=np.uint32)

# This is accepted
z |= y << 0x1 

# This throws an error
z |= x << 0x1
```
The bitshift is allowed, the bitwise or is only allowed with same numpy datatype:
```
TypeError: ufunc 'bitwise_or' output (typecode 'l') could not be coerced to provided output parameter (typecode 'I') according to the casting rule ''same_kind''
```

Reply:
What happens if you try
```suggestion
        column_value = np.array(column, dtype=np.uint32)
```